libbeat/outputs/shipper/shipper.go

```diff
@@ -164,90 +205,122 @@ func (c *shipper) Publish(ctx context.Context, batch publisher.Batch) error {
 	droppedCount := len(events) - len(nonDroppedEvents)
 
 	st.Dropped(droppedCount)
-	c.log.Debugf("%d events converted to protobuf, %d dropped", len(nonDroppedEvents), droppedCount)
-
-	ctx, cancel := context.WithTimeout(ctx, c.timeout)
-	defer cancel()
-	publishReply, err := c.client.PublishEvents(ctx, &messages.PublishRequest{
-		Uuid:   c.serverID,
-		Events: convertedEvents,
-	})
-
-	if status.Code(err) != codes.OK {
-		batch.Cancelled()         // does not decrease the TTL
-		st.Cancelled(len(events)) // we cancel the whole batch not just non-dropped events
-		return fmt.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(convertedEvents), err)
-	}
+	s.log.Debugf("%d events converted to protobuf, %d dropped", len(nonDroppedEvents), droppedCount)
 
-	// with a correct server implementation should never happen, this error is not recoverable
-	if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
-		return fmt.Errorf(
-			"server returned unexpected results, expected maximum accepted items %d, got %d",
-			len(nonDroppedEvents),
-			publishReply.AcceptedCount,
-		)
-	}
+	toSend := convertedEvents
+	var lastAcceptedIndex uint64
 
-	// so we explicitly close the stream
-	ackCtx, cancel := context.WithCancel(ctx)
+	ctx, cancel := context.WithTimeout(ctx, s.config.Timeout)
 	defer cancel()
-	ackClient, err := c.client.PersistedIndex(ackCtx, &messages.PersistedIndexRequest{
-		PollingInterval: durationpb.New(c.config.AckPollingInterval),
-	})
-	if err != nil {
-		return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
-	}
 
-	indexReply, err := ackClient.Recv()
-	if err != nil {
-		return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
-	}
+	for len(toSend) > 0 {
+		publishReply, err := s.client.PublishEvents(ctx, &messages.PublishRequest{
+			Uuid:   s.serverID,
+			Events: toSend,
+		})
 
-	if indexReply.GetUuid() != c.serverID {
-		batch.Cancelled()
-		st.Cancelled(len(events))
-		return fmt.Errorf("acknowledgement failed due to a connection to a different server %s, expected %s", indexReply.Uuid, c.serverID)
-	}
+		if status.Code(err) != codes.OK {
+			batch.Cancelled()         // does not decrease the TTL
+			st.Cancelled(len(events)) // we cancel the whole batch not just non-dropped events
+			return fmt.Errorf("failed to publish the batch to the shipper, none of the %d events were accepted: %w", len(convertedEvents), err)
+		}
 
-	for indexReply.PersistedIndex < publishReply.AcceptedIndex {
-		indexReply, err = ackClient.Recv()
-		if err != nil {
-			return fmt.Errorf("acknowledgement failed due to the connectivity error: %w", err)
+		// with a correct server implementation should never happen, this error is not recoverable
+		if int(publishReply.AcceptedCount) > len(nonDroppedEvents) {
```

Comment:
I might be missing something.

But I think we are sending len(toSend) events, so shouldn't we compare against that?  The first time, len(toSend) and len(nonDroppedEvents) should be equal but if we loop around wouldn't len(toSend) be less than len(nonDroppedEvents) ?

Reply:
@leehinman you're absolutely right, good catch, thanks!

Reply:
I fixed something including what you pointed out and also optimised memory usage of the whole function. It changed significantly in this PR and certain things were not up to date. See https://github.com/elastic/beats/pull/32785/commits/97422adabb01dad71497da6dab6ae5d4390b75eb